man/haiti.1
'\" t
.\"     Title: haiti
.\"    Author: Alexandre ZANNI (@noraj)
.\" Generator: Asciidoctor 2.0.23
.\"      Date: 2026-01-01
.\"    Manual: haiti manual
.\"    Source: haiti 4.0.0
.\"  Language: English
.\"
.TH "HAITI" "1" "2026-01-01" "haiti 4.0.0" "haiti manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE cyan R < >
.\}
.SH "NAME"
haiti \- CLI tool for identifying hash types (hash type identifier)
.SH "SYNOPSIS"
.sp
\fBhaiti\fP [\fIoptions\fP] \fI<hash>\fP
.sp
\fBhaiti\fP samples (\fI<ref>\fP | \fI<name>\fP)
.sp
\fBhaiti\fP [\fIoptions\fP] list
.sp
\fBhaiti\fP \-\-ascii\-art
.sp
\fBhaiti\fP \-h | \-\-help
.sp
\fBhaiti\fP \-\-version
.SH "OPTIONS"
.sp
\fB\-\-no\-color\fP
.RS 4
Disable colorized output (NO_COLOR environment variable is respected too).
.RE
.sp
\fB\-e, \-\-extended\fP
.RS 4
List all possible hash algorithms, including ones using salt.
.RE
.sp
\fB\-\-short\fP
.RS 4
Display in a short format: do not display hashcat and john the ripper references.
.RE
.sp
\fB\-\-hashcat\-only\fP
.RS 4
Show only hashcat references.
.RE
.sp
\fB\-\-john\-only\fP
.RS 4
Show only john the ripper references.
.RE
.sp
\fB\-\-ascii\-art\fP
.RS 4
Display the logo in colored ascii\-art.
.RE
.sp
\fB\-\-debug\fP
.RS 4
Display arguments.
.RE
.sp
\fB\-h, \-\-help\fP
.RS 4
Show this screen.
.RE
.sp
\fB\-\-version\fP
.RS 4
Show version.
.RE
.SH "COMMANDS"
.sp
\fBsamples\fP
.RS 4
Display hash samples for the given type.
.RE
.sp
\fBlist\fP
.RS 4
Display a list of all the available hash types.
.RE
.SH "PARAMETERS"
.sp
\fB<hash>\fP
.RS 4
Hash string to identify, read from STDIN if equal to "\-".
.RE
.sp
\fB<ref>\fP
.RS 4
Hashcat or John the Ripper reference.
.RE
.sp
\fB<name>\fP
.RS 4
Hash type name.
.RE
.SH "EXAMPLES"
.sp
\fBhaiti\fP \-e d41d8cd98f00b204e9800998ecf8427e
.sp
\fBhaiti\fP \-\-no\-color \-\-short d41d8cd98f00b204e9800998ecf8427e
.sp
b2sum /etc/os\-release | awk \*(Aq{print $1}\*(Aq | \fBhaiti\fP \-
.sp
\fBhaiti\fP samples crc32
.SH "COPYRIGHT"
.sp
Copyright © 2020\-2026 Alexandre ZANNI (independent)
.sp
Copyright © 2019\-2020 Alexandre ZANNI at Orange Cyberdefense
.sp
This software is distributed under the terms of the MIT License. You may freely
redistribute copies of it, subject to the conditions of the license.
.SH "VERSION"
.sp
4.0.0
.SH "REPORTING BUGS"
.sp
Report bugs to \c
.URL "https://github.com/noraj/haiti/issues" "" "."
.SH "RESOURCES"
.sp
Project homepage and documentation: \c
.URL "https://noraj.github.io/haiti" "" ""
.sp
Source code: \c
.URL "https://github.com/noraj/haiti" "" ""
.SH "SEE ALSO"
.sp
\fBhashcat\-haiti\fP(1), \fBjohn\-haiti\fP(1), \fBhaiti\-fzf\fP(1), \fBhaiti\-parsable\fP(1)
.SH "AUTHOR"
.sp
Alexandre ZANNI (@noraj)

docs/CHANGELOG.md
## [4.0.0]

- **Breaking changes**:
  - **Drop** Ruby 3.1 support, [EOL since 26/03/2025](https://www.ruby-lang.org/en/downloads/branches/) and no longer supported by minitest 6.x.x
- **Additions**:
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. / Argon2d KDF / without keyfile [#220][#220]
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. / Argon2d KDF / with keyfile
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. / Argon2id KDF / without keyfile
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. / Argon2id KDF / with keyfile
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. / AES-KDF / without keyfile
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. / AES-KDF / with keyfile
  - KeePass 2 (KDBX 3) / AES 256 enc. / AES-KDF / without keyfile
  - KeePass 2 (KDBX 3) / Twofish 256 enc. / AES-KDF / without keyfile
  - KeePass 2 (KDBX 3) / ChaCha20 256 enc. / AES-KDF / without keyfile
  - Apache Shiro 1 SHA-512 [#215][#215]
  - AS/400 DES [#222][#222]
  - AS/400 SSHA1 [#222][#222]
  - Blockchain, My Wallet, Legacy Wallets [#223][#223]
  - Blockchain, My Wallet, Second Password (SHA256) [#223][#223]
  - Blockchain, My Wallet, V2 [#223][#223]
  - Blockchain, My Wallet [#223][#223]
  - Cisco-ISE Hashed Password (SHA256) [#224][#224]
  - sm3crypt, SM3 (Unix) [#225][#225]
  - BLAKE2-256($salt.$pass) (blake2b) [#226][#226]
  - BLAKE2-256($pass.$salt) (blake2b) [#226][#226]
  - sha224(sha1($pass))
  - bcrypt(HMAC-SHA256($pass)) / bcryptsha256 v2 [#227][#227]
  - bcrypt(sha256($pass)) / bcryptsha256 [#227][#227]
  - md5($salt1.$pass.$salt2) [#227][#227]
  - md5($salt1.sha1($salt2.$pass)) [#227][#227]
  - md5($salt1.strtoupper(md5($salt2.$pass))) [#227][#227]
  - md5(md5($pass.$salt)) [#227][#227]
  - md5(md5($salt).md5(md5($pass))) [#227][#227]
  - md5(md5(md5($pass)).$salt) [#227][#227]
  - md5(md5(md5($pass.$salt1)).$salt2) [#227][#227]
  - md5(sha1($pass.$salt)) [#227][#227]
  - md5(sha1($salt.$pass)) [#227][#227]
  - md5(sha1(md5($pass))) [#227][#227]
  - sha256(sha256($pass.$salt)) [#227][#227]
  - sha512(sha512_bin($pass).$salt) [#227][#227]
- **Enhancements**:
  - OpenSSH private key
    - Fix Regexp for large key size (reported by [lap1nou][lap1nou] on [#217][#217])
  - BigCrypt
    - Better description
    - Add samples
    - Fix available JtR format
  - KeePass 2 (KDBX 4) / AES-Twofish-ChaCha20 256 enc. (Argon2d KDF & Argon2id KDF)
    - Add HC support
    - More precise regexp
  - SAP CODVN H (PWDSALTEDHASH) iSSHA-512
    - Add HC support
  - BLAKE2-256 (blake2b) [#226][#226]
    - Add HC support
  - MD6-256
    - Add HC support
  - sha224($pass.$salt)
    - Add HC support
    - Fix regexp for salt
  - sha224($salt.$pass)
    - Add HC support
  - sha224(sha224($pass))
    - Add HC support
  - md5($salt.md5($pass).$salt) [#227][#227]
    - Add HC support
    - Add sample
  - md5(md5(md5($pass).$salt1).$salt2) [#227][#227]
    - Add HC support
    - Add sample
  - sha512(sha512($pass).$salt) [#227][#227]
    - Add HC support
    - Add sample
- **Chore**:
  - Add support for Ruby 4.0

[#217]:https://github.com/noraj/haiti/pull/217
[lap1nou]:https://github.com/lap1nou
[#215]:https://github.com/noraj/haiti/issues/215
[#220]:https://github.com/noraj/haiti/issues/220
[#222]:https://github.com/noraj/haiti/issues/222
[#223]:https://github.com/noraj/haiti/issues/223
[#224]:https://github.com/noraj/haiti/issues/224
[#225]:https://github.com/noraj/haiti/issues/225
[#226]:https://github.com/noraj/haiti/issues/226
[#227]:https://github.com/noraj/haiti/issues/227

## [3.0.0]

- **Breaking changes**:
  - **Drop** Ruby 2.7 support, [EOL since 31/03/2023](https://www.ruby-lang.org/en/downloads/branches/), no longer compatible with newer versions of bundler
  - **Drop** Ruby 3.0 support (EOL)
- **Additions**:
  - WPA-PMKID-PMK [#162][#162]
  - WPA-EAPOL-PBKDF2 [#162][#162]
  - WPA-EAPOL-PMK [#162][#162]
  - WPA-PBKDF2-PMKID+EAPOL [#162][#162]
  - WPA-PMK-PMKID+EAPOL [#162][#162]
  - PFX (PKCS#12 PBE) (SHA1/SHA2) [#170][#170]
  - FileVault 2 [#216][#216]
- **Enhancements**:
  - OpenSSH private key
    - Better naming (add algorithm)
    - Better regexp (add john format support)
    - Add type `$2$`
  - IPB 2.x (Invision Power Board)
    - fix HC ref. [#182][#182] (thanks to @junewerner)
  - WPA-PMKID-PBKDF2
    - Different separators
- **Chore**:
  - Add support for Ruby 3.4
  - Change the way version is included. Instead of including a module `Version`, directly add the `VERSION` constant in the `HashIdentifier` class. It could avoid namespace collision with that kind of generic name.
- **Documentation**:
  - Use syntax highlight for shell commands without output
  - Add bash syntax highlight support
  - Update docsify, docsify plugins
  - Enhance of docsify theme
  - Include version number in docsify scripts and themes as well as CDN url for easier update
  - Comparison with more tools

[#182]:https://github.com/noraj/haiti/pull/182
[#162]:https://github.com/noraj/haiti/issues/162
[#170]:https://github.com/noraj/haiti/issues/170
[#216]:https://github.com/noraj/haiti/issues/216

## [2.1.0]

- **Additions**:
  - 4 new binaries
    - `hashcat-haiti`: wrapper for Hashcat where you can select the mode using haiti and fzf
    - `john-haiti`: wrapper for John the Ripper where you can select the format using haiti and fzf
    - `haiti-fzf`: select a Hashcat or John the Ripper reference with fzf from one of the matching hash types
      - mostly useful for `hashcat-haiti` and `john-haiti` or building another binary or alias that will make use of haiti with fzf input
    - `haiti-parsable`: display hash types matching that have a Hashcat reference in an easily parsable format
      - mostly useful for `haiti-fzf` or building another binary or alias
- **Documentation**:
  - migrating from ronn to asciidoc for man page generation
  - add documentation for [how to package haiti](https://noraj.github.io/haiti/#/pages/publishing?id=distribution-packaging) for distribution maintainers

lib/haiti.rb
# frozen_string_literal: true

# Ruby internal
require 'json'
# Project internal
require 'haiti/version'
require 'haiti/hash'

# The global Hash Identifier class
class HashIdentifier
  # Constants
  PROTOTYPES = JSON.parse(File.read(File.join(__dir__, '../data/prototypes.json')))
  COMMONS = JSON.parse(File.read(File.join(__dir__, '../data/commons.json')))

  # @return [String] the hash (as provided)
  # @example
  #   '5f4dcc3b5aa765d61d8327deb882cf99'
  attr_reader :hash

  # @return [Array<Chf>] list of {Chf} objects, representing the identified
  #   hashes
  attr_reader :type

  # A new instance of hash identifier
  # @param hash [String] the hash to identify
  def initialize(hash)
    @hash = hash
    @type = identify(hash)
    sort_commons
  end

  class << self
    # Find hash samples for a given hash type (by name or hashcat/john the ripper reference)
    # @param input [String|Integer] hash type name, hashcat reference, john the ripper reference
    # @return [Array<String>] a list of samples
    # @example
    #   HashIdentifier.samples('crc32')
    #   # => ["c762de4a:00000000", "$crc32$00000000.fa455f6b", "$crc32$4ff4f23f.ce6eb863", "5e23d60f:00000000"]
    def samples(input)
      samples = []
      PROTOTYPES.each do |prototype|
        prototype['modes'].each do |mode|
          if [mode['name'].downcase, mode['hashcat'].to_s,
              mode['john'].nil? ? mode['john'] : mode['john'].downcase].include?(input.to_s.downcase)
            samples << Chf.new(mode).samples
          end
        end
      end
      samples.delete(nil)
      samples.flatten.uniq
    end

    # List names of all hash types available
    # @return [Array<String>] a list of hash types name
    # @example
    #   HashIdentifier.list
    #   # => ["CRC-16", "CRC-16-CCITT", "FCS-16", "Adler-32", "CRC-32B", "FCS-32", ...]
    def list
      PROTOTYPES.flat_map { |d| d['modes'].map { |m| m['name'] } }.sort { |a, b| a.downcase <=> b.downcase }.uniq
    end

    # List all hash types available as <Chf> object
    # @return [Array<Chf>] a list of hash types object
    def object_list
      (PROTOTYPES.flat_map do |d|
         d['modes'].map do |m|
           Chf.new(m)
         end
       end).sort { |a, b| a.name.downcase <=> b.name.downcase }.uniq
    end
  end

  private

  # Check which hash types are matching the provided hash
  # @param hash [String] the hash to identify
  # @return [Array<Chf>] list of {Chf} objects, representing the identified
  #   hashes
  def identify(hash)
    res = []
    PROTOTYPES.each do |prototype|
      reg = Regexp.new prototype['regex'], Regexp::IGNORECASE
      next unless reg.match?(hash)

      prototype['modes'].each do |mode|
        res << Chf.new(mode)
      end
    end
    return res
  end

  # Sort common hash types first
  def sort_commons
    @type.sort_by! { |e| COMMONS.include?(e.name) ? 0 : 1 }
  end
end
